// @liaoliaots/nestjs-redis/dist/cluster/cluster.module.js
"use strict";
var ClusterModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClusterModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const cluster_manager_1 = require("./cluster-manager");
const cluster_providers_1 = require("./cluster.providers");
const cluster_constants_1 = require("./cluster.constants");
const common_2 = require("./common");
const messages_1 = require("../messages");
const utils_1 = require("../utils");
const cluster_logger_1 = require("./cluster-logger");
const errors_1 = require("../errors");
/**
 * @public
 */
let ClusterModule = ClusterModule_1 = class ClusterModule {
    constructor(options, clients) {
        this.options = options;
        this.clients = clients;
    }
    /**
     * Registers the module synchronously.
     */
    static forRoot(options, isGlobal = true) {
        const clusterClientProviders = (0, cluster_providers_1.createClusterClientProviders)();
        const providers = [
            (0, cluster_providers_1.createOptionsProvider)(options),
            cluster_providers_1.clusterClientsProvider,
            cluster_manager_1.ClusterManager,
            ...clusterClientProviders
        ];
        return {
            global: isGlobal,
            module: ClusterModule_1,
            providers,
            exports: [cluster_manager_1.ClusterManager, ...clusterClientProviders]
        };
    }
    /**
     * Registers the module asynchronously.
     */
    static forRootAsync(options, isGlobal = true) {
        var _a;
        if (!options.useFactory && !options.useClass && !options.useExisting) {
            throw new errors_1.MissingConfigurationError(messages_1.MISSING_CONFIGURATION);
        }
        const clusterClientProviders = (0, cluster_providers_1.createClusterClientProviders)();
        const providers = [
            ...(0, cluster_providers_1.createAsyncProviders)(options),
            cluster_providers_1.clusterClientsProvider,
            cluster_manager_1.ClusterManager,
            ...clusterClientProviders,
            ...((_a = options.extraProviders) !== null && _a !== void 0 ? _a : [])
        ];
        return {
            global: isGlobal,
            module: ClusterModule_1,
            imports: options.imports,
            providers,
            exports: [cluster_manager_1.ClusterManager, ...clusterClientProviders]
        };
    }
    async onApplicationShutdown() {
        if (this.options.closeClient) {
            const result = await (0, common_2.quitClients)(this.clients);
            result.forEach(([namespace, quit]) => {
                if ((0, utils_1.isResolution)(namespace) && (0, utils_1.isRejection)(quit) && (0, utils_1.isError)(quit.reason)) {
                    cluster_logger_1.logger.error(`${(0, utils_1.parseNamespace)(namespace.value)}: ${quit.reason.message}`);
                }
            });
        }
    }
};
ClusterModule = ClusterModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({}),
    tslib_1.__param(0, (0, common_1.Inject)(cluster_constants_1.CLUSTER_OPTIONS)),
    tslib_1.__param(1, (0, common_1.Inject)(cluster_constants_1.CLUSTER_CLIENTS)),
    tslib_1.__metadata("design:paramtypes", [Object, Object])
], ClusterModule);
exports.ClusterModule = ClusterModule;
