// @liaoliaots/nestjs-redis/dist/cluster/cluster.providers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClusterClientProviders = exports.clusterClientsProvider = exports.createAsyncOptionsProvider = exports.createAsyncOptions = exports.createAsyncProviders = exports.createOptionsProvider = void 0;
const cluster_constants_1 = require("./cluster.constants");
const common_1 = require("./common");
const cluster_manager_1 = require("./cluster-manager");
const default_options_1 = require("./default-options");
const createOptionsProvider = (options) => ({
    provide: cluster_constants_1.CLUSTER_OPTIONS,
    useValue: Object.assign(Object.assign({}, default_options_1.defaultClusterModuleOptions), options)
});
exports.createOptionsProvider = createOptionsProvider;
const createAsyncProviders = (options) => {
    if (options.useFactory) {
        return [
            {
                provide: cluster_constants_1.CLUSTER_OPTIONS,
                useFactory(options) {
                    return Object.assign(Object.assign({}, default_options_1.defaultClusterModuleOptions), options);
                },
                inject: [cluster_constants_1.CLUSTER_INTERNAL_OPTIONS]
            },
            (0, exports.createAsyncOptionsProvider)(options)
        ];
    }
    if (options.useClass) {
        return [
            {
                provide: options.useClass,
                useClass: options.useClass
            },
            (0, exports.createAsyncOptionsProvider)(options)
        ];
    }
    if (options.useExisting)
        return [(0, exports.createAsyncOptionsProvider)(options)];
    return [];
};
exports.createAsyncProviders = createAsyncProviders;
const createAsyncOptions = async (optionsFactory) => {
    const options = await optionsFactory.createClusterOptions();
    return Object.assign(Object.assign({}, default_options_1.defaultClusterModuleOptions), options);
};
exports.createAsyncOptions = createAsyncOptions;
const createAsyncOptionsProvider = (options) => {
    if (options.useFactory) {
        return {
            provide: cluster_constants_1.CLUSTER_INTERNAL_OPTIONS,
            useFactory: options.useFactory,
            inject: options.inject
        };
    }
    if (options.useClass) {
        return {
            provide: cluster_constants_1.CLUSTER_OPTIONS,
            useFactory: exports.createAsyncOptions,
            inject: [options.useClass]
        };
    }
    if (options.useExisting) {
        return {
            provide: cluster_constants_1.CLUSTER_OPTIONS,
            useFactory: exports.createAsyncOptions,
            inject: [options.useExisting]
        };
    }
    return {
        provide: cluster_constants_1.CLUSTER_OPTIONS,
        useValue: {}
    };
};
exports.createAsyncOptionsProvider = createAsyncOptionsProvider;
exports.clusterClientsProvider = {
    provide: cluster_constants_1.CLUSTER_CLIENTS,
    useFactory: (options) => {
        var _a;
        const clients = new Map();
        if (Array.isArray(options.config) /* multiple */) {
            options.config.forEach(item => { var _a; return clients.set((_a = item.namespace) !== null && _a !== void 0 ? _a : cluster_constants_1.DEFAULT_CLUSTER_NAMESPACE, (0, common_1.createClient)(item)); });
        }
        else if (options.config /* single */) {
            clients.set((_a = options.config.namespace) !== null && _a !== void 0 ? _a : cluster_constants_1.DEFAULT_CLUSTER_NAMESPACE, (0, common_1.createClient)(options.config));
        }
        if (options.readyLog)
            (0, common_1.displayReadyLog)(clients);
        if (options.errorLog)
            (0, common_1.displayErrorLog)(clients);
        return clients;
    },
    inject: [cluster_constants_1.CLUSTER_OPTIONS]
};
const createClusterClientProviders = () => {
    const providers = [];
    common_1.namespaces.forEach((token, namespace) => {
        providers.push({
            provide: token,
            useFactory: (clusterManager) => clusterManager.getClient(namespace),
            inject: [cluster_manager_1.ClusterManager]
        });
    });
    return providers;
};
exports.createClusterClientProviders = createClusterClientProviders;
