// @liaoliaots/nestjs-redis/dist/utils/is.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNullish = exports.isDirectInstanceOf = exports.isRejection = exports.isResolution = exports.isError = exports.isSymbol = exports.isString = void 0;
/**
 * Returns `true` if the value is of type `string`.
 *
 * @param value - Any value
 */
const isString = (value) => typeof value === 'string';
exports.isString = isString;
/**
 * Returns `true` if the value is of type `symbol`.
 *
 * @param value - Any value
 */
const isSymbol = (value) => typeof value === 'symbol';
exports.isSymbol = isSymbol;
/**
 * Returns `true` if the value is an `Error`.
 *
 * @param value - Any value
 */
const isError = (value) => {
    // eslint-disable-next-line @typescript-eslint/unbound-method
    const { toString } = Object.prototype;
    const typeName = toString.call(value).slice(8, -1);
    return typeName === 'Error';
};
exports.isError = isError;
/**
 * Returns `true` if the value is of type `PromiseFulfilledResult`.
 *
 * @param value - PromiseSettledResult
 */
const isResolution = (value) => {
    return value.status === 'fulfilled';
};
exports.isResolution = isResolution;
/**
 * Returns `true` if the value is of type `PromiseRejectedResult`.
 *
 * @param value - PromiseSettledResult
 */
const isRejection = (value) => {
    return value.status === 'rejected';
};
exports.isRejection = isRejection;
/**
 * Returns `true` if the value is a direct instance of `class`.
 *
 * @param value - Any value
 * @param class_ - Any class
 */
const isDirectInstanceOf = (value, class_) => {
    return Object.getPrototypeOf(value) === class_.prototype;
};
exports.isDirectInstanceOf = isDirectInstanceOf;
/**
 * Returns `true` if the value is not defined (=== null, === undefined).
 *
 * @param value - Any value
 */
const isNullish = (value) => value === undefined || value === null;
exports.isNullish = isNullish;
