// @cherie-xf/nestjs-microservice-otel/dist/OpenTelemetryModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenTelemetryModule = void 0;
const sdk_node_1 = require("@opentelemetry/sdk-node");
const TraceService_1 = require("./Trace/TraceService");
const TraceClientProxy_1 = require("./Trace/TraceClientProxy");
const Constants_1 = require("./Constants");
const MetricService_1 = require("./Metric/MetricService");
const OpenTelemetryModuleConfig_1 = require("./OpenTelemetryModuleConfig");
const OpenTelemetryService_1 = require("./OpenTelemetryService");
const DecoratorInjector_1 = require("./Trace/Injectors/DecoratorInjector");
const core_1 = require("@nestjs/core");
const MetricHttpMiddleware_1 = require("./Metric/Interceptors/Http/MetricHttpMiddleware");
const MetricInterceptor_1 = require("./Metric/Interceptors/MetricInterceptor");
const event_emitter_1 = require("@nestjs/event-emitter");
const MetricHttpEventProducer_1 = require("./Metric/Interceptors/Http/MetricHttpEventProducer");
const MetricGrpcEventProducer_1 = require("./Metric/Interceptors/Grpc/MetricGrpcEventProducer");
const MetricRabbitMQEventProducer_1 = require("./Metric/Interceptors/RabbitMQ/MetricRabbitMQEventProducer");
const DecoratorObserverMetricInjector_1 = require("./Metric/Injectors/DecoratorObserverMetricInjector");
const DecoratorCounterMetricInjector_1 = require("./Metric/Injectors/DecoratorCounterMetricInjector");
const sdk_metrics_base_1 = require("@opentelemetry/sdk-metrics-base");
const sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
const resources_1 = require("@opentelemetry/resources");
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const sdk_trace_node_1 = require("@opentelemetry/sdk-trace-node");
const instrumentation_http_1 = require("@opentelemetry/instrumentation-http");
const instrumentation_1 = require("@opentelemetry/instrumentation");
const instrumentation_nestjs_core_1 = require("@opentelemetry/instrumentation-nestjs-core");
const RpcTraceContextPropagator_1 = require("./Trace/RpcTraceContextPropagator");
class OpenTelemetryModule {
    configure(consumer) {
        consumer.apply(MetricHttpMiddleware_1.MetricHttpMiddleware).forRoutes('*');
    }
    static async forRoot(configuration = {}) {
        configuration = { ...OpenTelemetryModuleConfig_1.OpenTelemetryModuleDefaultConfig, ...configuration };
        const injectors = configuration?.traceAutoInjectors ?? [];
        const metrics = configuration?.metricAutoObservers ?? [];
        return {
            global: true,
            module: OpenTelemetryModule,
            imports: [event_emitter_1.EventEmitterModule.forRoot()],
            providers: [
                ...injectors,
                ...metrics,
                TraceService_1.TraceService,
                TraceClientProxy_1.TraceClientProxy,
                MetricService_1.MetricService,
                OpenTelemetryService_1.OpenTelemetryService,
                MetricHttpMiddleware_1.MetricHttpMiddleware,
                MetricHttpEventProducer_1.MetricHttpEventProducer,
                MetricGrpcEventProducer_1.MetricGrpcEventProducer,
                MetricRabbitMQEventProducer_1.MetricRabbitMQEventProducer,
                DecoratorInjector_1.DecoratorInjector,
                DecoratorObserverMetricInjector_1.DecoratorObserverMetricInjector,
                DecoratorCounterMetricInjector_1.DecoratorCounterMetricInjector,
                this.buildProvider(configuration),
                this.buildInjectors(configuration),
                this.buildTracerProvider(configuration),
                this.buildMeter(),
                this.buildTracer(),
                {
                    provide: Constants_1.Constants.SDK_CONFIG,
                    useValue: configuration,
                },
                {
                    provide: core_1.APP_INTERCEPTOR,
                    useClass: MetricInterceptor_1.MetricInterceptor,
                },
            ],
            exports: [
                TraceService_1.TraceService,
                TraceClientProxy_1.TraceClientProxy,
                MetricService_1.MetricService,
                sdk_metrics_base_1.Meter,
                sdk_trace_base_1.Tracer,
                sdk_trace_node_1.NodeTracerProvider,
            ],
        };
    }
    static buildTracerProvider(configuration) {
        return {
            provide: sdk_trace_node_1.NodeTracerProvider,
            useFactory: async () => {
                if (configuration.serviceName) {
                    configuration.resource = new resources_1.Resource({
                        [semantic_conventions_1.SemanticResourceAttributes.SERVICE_NAME]: configuration.serviceName,
                    });
                }
                const provider = new sdk_trace_node_1.NodeTracerProvider(configuration);
                provider.addSpanProcessor(configuration.spanProcessor);
                provider.register({ propagator: new RpcTraceContextPropagator_1.RpcTraceContextPropagator() });
                (0, instrumentation_1.registerInstrumentations)({
                    instrumentations: [
                        new instrumentation_http_1.HttpInstrumentation(),
                        new instrumentation_nestjs_core_1.NestInstrumentation(),
                    ],
                });
                return provider;
            },
        };
    }
    static buildProvider(configuration) {
        return {
            provide: Constants_1.Constants.SDK,
            useFactory: async () => {
                const sdk = new sdk_node_1.NodeSDK(configuration);
                await sdk.start();
                return sdk;
            },
        };
    }
    static buildInjectors(configuration) {
        const injectors = configuration?.traceAutoInjectors ?? [];
        const metrics = configuration?.metricAutoObservers ?? [];
        return {
            provide: Constants_1.Constants.SDK_INJECTORS,
            useFactory: async (...injectors) => {
                for await (const injector of injectors) {
                    if (injector['inject'])
                        await injector.inject();
                }
            },
            inject: [
                DecoratorInjector_1.DecoratorInjector,
                DecoratorObserverMetricInjector_1.DecoratorObserverMetricInjector,
                DecoratorCounterMetricInjector_1.DecoratorCounterMetricInjector,
                ...injectors,
                ...metrics,
            ],
        };
    }
    static async forRootAsync(configuration = {}) {
        return {
            global: true,
            module: OpenTelemetryModule,
            imports: [...configuration?.imports, event_emitter_1.EventEmitterModule.forRoot()],
            providers: [
                TraceService_1.TraceService,
                TraceClientProxy_1.TraceClientProxy,
                MetricService_1.MetricService,
                OpenTelemetryService_1.OpenTelemetryService,
                MetricHttpMiddleware_1.MetricHttpMiddleware,
                MetricHttpEventProducer_1.MetricHttpEventProducer,
                MetricGrpcEventProducer_1.MetricGrpcEventProducer,
                MetricRabbitMQEventProducer_1.MetricRabbitMQEventProducer,
                this.buildAsyncProvider(),
                this.buildAsyncInjectors(),
                this.buildAsyncTraceProvider(),
                this.buildMeter(),
                this.buildTracer(),
                {
                    provide: Constants_1.Constants.SDK_CONFIG,
                    useFactory: configuration.useFactory,
                    inject: configuration.inject,
                },
                {
                    provide: core_1.APP_INTERCEPTOR,
                    useClass: MetricInterceptor_1.MetricInterceptor,
                },
            ],
            exports: [TraceService_1.TraceService, TraceClientProxy_1.TraceClientProxy, MetricService_1.MetricService, sdk_metrics_base_1.Meter, sdk_trace_base_1.Tracer],
        };
    }
    static buildAsyncTraceProvider() {
        return {
            provide: sdk_trace_node_1.NodeTracerProvider,
            useFactory: async (config) => {
                if (config.serviceName) {
                    config.resource = new resources_1.Resource({
                        [semantic_conventions_1.SemanticResourceAttributes.SERVICE_NAME]: config.serviceName,
                    });
                }
                const provider = new sdk_trace_node_1.NodeTracerProvider(config);
                provider.addSpanProcessor(config.spanProcessor);
                provider.register({ propagator: new RpcTraceContextPropagator_1.RpcTraceContextPropagator() });
                provider.addSpanProcessor(config.spanProcessor);
                (0, instrumentation_1.registerInstrumentations)({
                    instrumentations: [
                        new instrumentation_http_1.HttpInstrumentation(),
                        new instrumentation_nestjs_core_1.NestInstrumentation(),
                    ],
                });
                provider.register();
                return provider;
            },
        };
    }
    static buildAsyncProvider() {
        return {
            provide: Constants_1.Constants.SDK,
            useFactory: async (config) => {
                const sdk = new sdk_node_1.NodeSDK(config);
                await sdk.start();
                return sdk;
            },
            inject: [Constants_1.Constants.SDK_CONFIG],
        };
    }
    static buildAsyncInjectors() {
        return {
            provide: Constants_1.Constants.SDK_INJECTORS,
            useFactory: async (config, moduleRef) => {
                config = { ...OpenTelemetryModuleConfig_1.OpenTelemetryModuleDefaultConfig, ...config };
                const injectors = config.traceAutoInjectors ??
                    OpenTelemetryModuleConfig_1.OpenTelemetryModuleDefaultConfig.traceAutoInjectors;
                const metrics = config.metricAutoObservers ??
                    OpenTelemetryModuleConfig_1.OpenTelemetryModuleDefaultConfig.metricAutoObservers;
                const decoratorInjector = await moduleRef.create(DecoratorInjector_1.DecoratorInjector);
                await decoratorInjector.inject();
                const decoratorObserverMetricInjector = await moduleRef.create(DecoratorObserverMetricInjector_1.DecoratorObserverMetricInjector);
                await decoratorObserverMetricInjector.inject();
                const decoratorCounterMetricInjector = await moduleRef.create(DecoratorCounterMetricInjector_1.DecoratorCounterMetricInjector);
                await decoratorCounterMetricInjector.inject();
                for await (const injector of injectors) {
                    const created = await moduleRef.create(injector);
                    if (created['inject'])
                        await created.inject();
                }
                for await (const metric of metrics) {
                    const createdMetric = await moduleRef.create(metric);
                    if (createdMetric['inject'])
                        await createdMetric.inject();
                }
                return {};
            },
            inject: [Constants_1.Constants.SDK_CONFIG, core_1.ModuleRef],
        };
    }
    static buildMeter() {
        return {
            provide: sdk_metrics_base_1.Meter,
            useFactory: (metricService) => metricService.getMeter(),
            inject: [MetricService_1.MetricService],
        };
    }
    static buildTracer() {
        return {
            provide: sdk_trace_base_1.Tracer,
            useFactory: (traceService) => traceService.getTracer(),
            inject: [TraceService_1.TraceService],
        };
    }
}
exports.OpenTelemetryModule = OpenTelemetryModule;
